src/livestatus.c

```diff
@@ -0,0 +1,479 @@
+      return -1;
+    }
+
+    memset(&sun, 0x0, sizeof(sun));
+    sun.sun_family = AF_UNIX;
+    strcpy(sun.sun_path, sockfile);
+
+    do {
+      rc = connect(sfd, (struct sockaddr *)&sun, sizeof(sun));
+    } while (errno == EINTR);
+
+    if (rc < 0) {
+      ERROR("livestatus plugin: connect to socket file %s: %s", sockfile,
+            STRERRNO);
+      return -1;
+    }
+
+    *sockfd = sfd;
+
+    return 0;
+  } /* unix_connect */
+
+  static int ls_parse(const char *lresponse, livestatus_status_t *lstatus) {
+    int rc = -1;
+    int i = 0;
+    char **fields = NULL;
+    char *pchar = NULL;
+
+    fields = malloc(sizeof(char **) * LIVESTATUS_EXPECTED_FIELDS_RESP_NB);
+    if (fields == NULL) {
+      ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+      return -1;
+    }
+
+    for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+      fields[j] = calloc(1, sizeof(char *) * 32);
+      if (fields[j] == NULL) {
+        ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+        rc = -1;
+        goto free_all_fields;
+      }
+    }
+
+    pchar = (char *)lresponse;
+    char *wstart = (char *)lresponse;
+    int exit = 0;
+    while (!exit) {
+      if (*pchar == '\0' || *pchar == '\n') {
+        exit = 1;
+      } else if (*pchar != LIVESTATUS_FIELD_SEP) {
+        pchar++;
+        continue;
+      }
+
+      strncpy(fields[i], wstart, pchar - wstart);
+      wstart = ++pchar;
+      i++;
+
+      if (!exit) {
+        if (i >= LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+          ERROR("livestatus plugin: too many fields in livestatus output");
+          rc = -2;
+          goto free_all_fields;
+        }
+      }
+    }
+
+    if (i < LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+      ERROR("livestatus plugin: not enough fields in livestatus output");
+      rc = -3;
+      goto free_all_fields;
+    }
+
+    *lstatus = c_to_livestatus_status((const char **)fields);
+
+    rc = 0;
+
+  free_all_fields:
+    for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+      if (fields[j] != NULL) {
+        free(fields[j]);
+        fields[j] = NULL;
+      }
+    }
+
+    free(fields);
+    fields = NULL;
+
+    return rc;
+  } /* int ls_parse */
+
+  static int ls_read_parse(const int sockfd, livestatus_status_t *lstatus) {
+    ssize_t bread = -1;
+    char buffer[IO_BUFFER_SIZE];
+
+    do {
+      memset(buffer, 0x0, sizeof(buffer));
+      bread = read((int)sockfd, buffer, IO_BUFFER_SIZE - 1);
+    } while (errno == EINTR);
+
+    if (bread < 0) {
+      ERROR("livestatus plugin: reading from socket: %s", STRERRNO);
+      return -1;
+    }
+
+    buffer[IO_BUFFER_SIZE - 1] = '\0';
+
+    return ls_parse((const char *)buffer, lstatus);
+  } /* int ls_read_parse */
+
+  static int ls_send_request(const int sockfd) {
+    int written = -1;
+    char request[512];
+
+    memset(request, 0x0, sizeof(request));
+
+    snprintf(request, sizeof(request) - 1, "GET status\nColumns: %s\n\n",
```

Comment:
Please use `ssnprintf` (note the extra "s") instead. *sprintf(3)* may not null-terminate the result. This is especially problematic when using the `%s` format.

Reply:
Hi @octo . I do not have any problem using `ssnprintf()` here.
But I'm wondering if there is a real advantage over the `snprintf()` function I'm using here ?
A quick look at the `ssnprintf()` function shows me that this is a simple wrapper on top of `vsnprintf()`. What would be the benefit ?

Reply:
We introduced `ssnprintf` and `sstrncpy` to ensure that the result is null terminated. Later, we realized that `snprintf` already does null termination. That's probably why the function doesn't do anything different from vanilla `snprintf` and should be removed.

So with that in mind, never mind, feel free to use `snprintf` directly. Sorry for the noise ;)